### ISSUE TITLE
Payment is immediately being canceled due to ErrorCode: 2850

### ISSUE THREAD

**Issue by DanielCauser** [stripe/stripe-terminal-ios#112]
## Summary
We're having an issue where the first transaction of the day immediately cancels due to error 2850. This is awful, especially because even retrying the same transaction, they get the same error, the only way to make it work is to actually cancel the payment and try again, which puts the client funds on hold.

What we do to reproduce:
1. Use the app normally for one day.
2. On the next day do (With the same app, or a fresh install)
3. Connect iOS app to the Bluetooth terminal (SDK connect).
4. Create a payment intent.
5. Collect Payment.
6. Payment cancels as soon as the user inserts the card in the terminal with error 2850. Even though there are no cards on the card reader.

![image](https://user-images.githubusercontent.com/7211075/133143954-ca346160-0fa7-4fb7-8f54-406dc92d0d20.png)

Is there any suggestion I can do to try to troubleshoot this problem?

## Code to reproduce
Payment service:
```
//Step 1 - Create a PaymentIntent
var paymentIntent = await _paymentService.CreatePaymentIntent(ct, parameters);

//Save the PaymentIntent in case we need to cancel it.
_currentIntent = paymentIntent;

//Step 2 - Collect payment method and update the PaymentIntent
paymentIntent = await _paymentService.CollectPaymentMethod(ct, paymentIntent, statusMessageCallback);
```

Within ` _paymentService.CollectPaymentMethod`
```
var tcs = new TaskCompletionSource<SCPPaymentIntent>();
var displayDelegate = new StripePaymentServiceDelegates.ReaderDisplayDelegate(readerDisplayCallback, _resourcesService);

using (await _transactionLock.LockAsync(ct))
{
_collectPaymentMethodCancelable = SCPTerminal.Shared.CollectPaymentMethod(intent.Intent, displayDelegate, (paymentIntent, error) =>
{
	if (paymentIntent != null)
	{
		tcs.TrySetResult(paymentIntent);
	}
	else
	{
		SetExceptionFromStripeError(tcs, error, "CollectPaymentMethod failed");
	}
});
}
```

## iOS version
12+

## Installation method
N/A

## SDK version
SDK Version is 1.3.0

**Comment by bric-stripe** [comment]
👋 hi, we've been getting some extra details and debugging through an existing support thread but wanted to chime in here that we've been able to identify a race condition where if the card is inserted very close to the same time the `collectPaymentMethod` method is called. In my testing, it's within ~500ms[0] of calling `collectPaymentMethod` and the `CardLeftInReader` error _may_ occur. So far this is the only case we've able to reproduce this and apologies that it took a good bit of time to find this.

We'll be working to fix the race condition, but for a short term the recommendation is to make sure the customer is not instructed to insert/swipe/tap _until_ the Terminal SDK has delivered the `didRequestReaderInput:` (either to the [display delegate in 1.x](https://stripe.dev/stripe-terminal-ios/1.4.0/Protocols/SCPReaderDisplayDelegate.html) or to the [bluetooth reader delegate in 2.x](https://stripe.dev/stripe-terminal-ios/docs/Protocols/SCPBluetoothReaderDelegate.html#/c:objc(pl)SCPBluetoothReaderDelegate(im)reader:didRequestReaderInput:). At the point that is called the reader is confidently awaiting payment and it should avoid the race condition.

[0] ~500ms is _not_ accurate, but is the best estimate I could come up with. The actual time required may vary a good bit but waiting for the delegate method should always be reliable.

**Comment by DanielCauser** [comment]
Amazing insight, thank you! Since you recommend that we wait for the `didRequestReaderInput` delegate to allow the user to insert the card, I wonder if there is a specific delegate that is invoked when we want the user to **Remove the card** from the card reader? I want to make sure that we also "force", the user to remove the card before we can move on to _Process the payment_

**Comment by DanielCauser** [comment]
Hey @bric-stripe, your first suggestion works like a charm!

Regarding the suggestion on waiting for the user to remove the card, I have one hiccup.  `didReportReaderEvent:info:` only gets invoked if the user actually inserts/removes the card! 

If the user just swipes or taps the card, the `didReportReaderEvent:info:` is never invoked, so I can't really tell when to let the user move on. Is there an object in the terminal, or collectPaymentMethod class that could tell me what action the user took? Etiher Swipe/Tap/Insert the card in the card reader?

**Comment by bric-stripe** [comment]
> If I could match the information of what type of action took with the card, it would be perfect

hrmmm, I don't know that we have great options here today. 2 thoughts:
1. only wait for CardRemoved if CardInserted was received...otherwise assume swipe or tap/mobile wallet and move on after collect's completion block is called
2. if you call process and _then_ add the waiting you can check the charge's [read_method](https://stripe.com/docs/api/charges/object#charge_object-payment_method_details-card_present-read_method) on the `PaymentIntent` returned in process. It's a little clumsy to get at though and I realize this isn't the point at which you want to know the read method (since this is only on the payment intent returned from process, not capture), so may not be helpful:

```
(lldb) po processedIntent?.charges.first?.paymentMethodDetails?.cardPresent?.originalJSON["read_method"]
▿ Optional<Any>
  - some : contactless_emv
```

**Comment by bric-stripe** [comment]
Hi, I'm closing out stale issues and realized this one still has some loose ends, but I think the main issue should be fixed. Specifically there was a fix for CardLeftInReader detection in 2.9.1 and the latest SDK version should now reliable detect card left in reader when the next transaction starts and I believe the card inserted and card removed `didReportReaderEvent` should be able to be reliably used.

The loose end though is the SDK still doesn't report card tapped/swiped during collect and that is a reasonable request that we are tracking in an internal ticket but no timeline for when it would be available.

One new option to consider is the [`updatePaymentIntent`](https://stripe.dev/stripe-terminal-ios/docs/Classes/SCPCollectConfiguration.html#/c:objc(cs)SCPCollectConfiguration(py)updatePaymentIntent) flag in the `CollectConfiguration` is now available and would let you identiify the `read_method` after collect, but at the expense of a network call during collect.
